File: 0083-remove-duplicates-from-sorted-list/0083-remove-duplicates-from-sorted-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* deleteDuplicates(ListNode* head) {
        unordered_map<int, int> listMap;
        ListNode* temp = nullptr;
        ListNode* list = head;
        ListNode* prev = list;

        ListNode* listHead = head;

        while (list!=nullptr) {

            if (listMap.find(list->val) == listMap.end()) {
                listMap[list->val] = list->val;
            } else {
                temp = list;
                prev->next = list->next;
                //delete(temp);
                //temp = NULL;
                //list = temp;
                list = prev;
            }
            prev = list;
            list = list->next;
        }

        return listHead;
    }
};

File: 0083-remove-duplicates-from-sorted-list/0083-remove-duplicates-from-sorted-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def deleteDuplicates(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if head == None or head.next == None:
            return head
        prev: ListNode = None
        temp: ListNode = head
        while temp != None:
            if prev == None:
                prev = temp
            elif prev.val == temp.val:
                # bad stuff
                prev.next = temp.next
                #prev = prev.next
                #temp = temp.next
            else:
                prev = temp
            temp = temp.next
        return head
